drivers/place/attendee_scanner.cr

```diff
@@ -0,0 +1,88 @@
+require "placeos-driver"
+require "json"
+
+class Place::AttendeeScanner < PlaceOS::Driver
+  descriptive_name "PlaceOS Room Events"
+  generic_name :AttendeeScanner
+
+  accessor staff_api : StaffAPI_1
+
+  default_settings({
+    internal_domains: [] of String,
+  })
+
+  getter systems : Hash(String, Array(String)) { get_systems_list.not_nil! }
+  getter internal_domains : Array(String) = [] of String
+
+  def on_update
+    @internal_domains = setting(Array(String), :internal_domains)
+  end
+
+  # Grabs the list of systems in the building
+  def get_systems_list
+    building_id = system[:Location].building_id.get
+    system["StaffAPI"].systems_in_building(building_id).get.as_h.transform_values(&.as_a.map(&.as_s))
+  rescue error
+    logger.warn(exception: error) { "unable to obtain list of systems in the building" }
+    nil
+  end
+
+  # Lists external guests based on email domains
+  def list_external_guests
+    meetings = [] of Hash(String, JSON::Any)
+
+    systems.each do |level_id, system_ids|
+      system_ids.each do |system_id|
+        period_start = Time.local.to_unix
+        period_end = Time.local.to_unix + 86400
+
+        events = staff_api.query_events(period_start, period_end, systems: [system_id]).get
+
+        events.each do |event|
+          domain = event.as_h.["host"].as_s.split("@").last
+
+          unless internal_domains.includes?(domain)
+            event_id = event.as_h.["id"].as_s
+            ical_uid = event.as_h.["ical_uid"].as_s
+
+            bookings = staff_api.query_bookings(event_id: event_id, ical_uid: ical_uid).get
+
+            attendee_emails = event["attendees"].as_a.map do |attendee|
+              attendee.as_h.["email"].as_s
+            end
+
+            bookings.as_a.each do |booking|
+              booking.as_h.["attendees"].as_a.each do |attendee|
+                unless attendee_emails.includes?(attendee.as_h.["email"].as_s)
+                  staff_api.create_booking(
```

Comment:
Property worth asking @MrYuion what he includes when creating visitor bookings from the frontend as it should be the same.

Reply:
@MrYuion please provide some data

Reply:
This is what I'd send normally. I send my whole local model though, so you probably don't need all of the fields.
```json
{
    "asset_id": "<visitor_email>",
    "asset_ids": ["<visitor_email>"],
    "asset_name": "alex",
    "zones": ["<org_zone_id>", "<region_zone_id>", "<building_zone_id>"],
    "booking_start": 1727847000,
    "booking_end": 1727848800,
    "booking_type": "visitor",
    "type": "visitor",
    "timezone": "Australia/Sydney",
    "user_email": "<host_email>",
    "title": "alex",
    "description": "<event_title>",
    "checked_in": false,
    "rejected": false,
    "approved": false,
    "deleted": false,
    "extension_data": {
        "parent_id": "<event_id>",
        "name": "alex",
        "location_id": "<resource_name>",
        "assets": [],
        "tags": []
    },
    "access": false,
    "permission": "PRIVATE",
    "attendees": [
        {
            "id": "<visitor_email>",
            "name": "alex",
            "email": "<visitor_email>",
            "first_name": "alex",
            "last_name": "",
            "phone": "",
            "organisation": "",
            "notes": "",
            "photo": "",
            "username": "",
            "organizer": false,
            "checked_in": false,
            "required": true,
            "resource": false,
            "response_status": "",
            "department": "",
            "groups": [],
            "extension_data": {},
            "is_external": true,
            "visit_expected": true,
            "assistance_required": false
        }
    ],
    "tags": [],
    "images": [],
    "all_day": false,
    "linked_bookings": [],
    "status": "tentative",
    "process_state": "pending",
    "recurrence_type": "none"
}
```
With the `event_id` and `ical_uid` as query parameters.
`/api/staff/v1/bookings?ical_uid=<event_ical_uid>&event_id=<event_id>`